teps/0042-taskrun-breakpoint-on-failure.md

```diff
@@ -0,0 +1,113 @@
+
+### Goals
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  generateName: quatredeux
+spec:
+  breakpoint:
+    onFailure: true
+```
+
+When the "breakpoint on failure" spec is mentioned in a particular TaskRun
+
+- pause TaskRun on failure of a step
+- failed step should not exit
+- user should be able to get remote shell access to the step container to analyze and debug
+
+### Non-Goals
+
+The scope of this TEP does not exceed any other kind of breakpointing or debugging apart from breakpointing on failure and failure itself.
+
+### Use Cases (optional)
+
+This enhancement can allow a more interactive way of working with TaskRuns. CLI or IDE extensions could allow users to run a 
+Task with breakpoint-on-failure and while the logs are being streamed, the user can be dropped to a shell in the breakpoint
+failed step and allow the user to debug as they develop.
+
+## Requirements
+
+* To enable breakpoint on failure of a step, it would be necessary to update the Entrypoint to support lifecycle
+changes when breakpoint on failure has been enabled.
+
+* The step environment would have to have scripts which the user can use to clear the breakpoint or
+mark the step as a success or a failure.
+
+* The entrypoint needs information about it's position in to the order of execution of steps. This will allow the
+scripts to know for sure what kind of changes need to be made to manipulate the lifecycle of the step.
+
+## Proposal
+
+TaskRuns will add a new spec called `breakpoint` under which `onFailure` can be set to true to enable breakpointing
+on failure.
+
+### Controlling step lifecycle
+
+Currently, the entrypoint writes the err postfile if there was an error during waiting for the previous postfile. This would
+be disabled so that the user who is debugging the step can control the lifecycle of a step and mark it either 
+as a success or failure by themselves.
+
+This can further be simplified for the user by providing scripts in the environment which the user can run.
+
+### Exiting breakpoint
+
+To exit a step which has been paused upon failure, the step would wait on a file similar to `<step-no>.breakpointexit` which 
+would unpause and exit the step container. eg: Step 0 fails and is paused. Writing `0.breakpointexit` in `/tekton/tools`
+would unpause and exit the step container.
+
+### Step information 
+
+For the step to have an understanding of where it stands in the order of execution, we can provde a mount which can be
+read by the entrypoint. More information below.
+
+### Mounts
+
+`/tekton/debug/scripts` : Contains scripts which the user can run to mark the step as a success, failure or exit the breakpoint.
+Shared between all the containers.
+
+`/tekton/debug/info/<n>` : Contains information about the step. Single EmptyDir shared between all step containers, but renamed 
+to reflect step number. eg: Step 0 will have `/tekton/debug/info/0`, Step 1 will have `/tekton/debug/info/1` etc.
```

Comment:
What information do we see would be available here ? The `entrypoint` process can already infer where it is in the order of execution (from flags / files it listens to). What other kind of informations do we need / want ?

Reply:
You are right, info in this case is no longer needed. As we can pick it up from the postfile.

Reply:
But for the scirpts to know which step number it is running in to write the correct file, this would be necessary.